### ISSUE TITLE
Support Firefox as a browser in flutter run

### ISSUE THREAD

**Issue by jonahwilliams** [flutter/flutter#55324]
The Firefox remote protocol claims to support a subset of the chrome devtools protocol: https://firefox-source-docs.mozilla.org/remote/index.html . If this subset is sufficient, we might be able to connect with out existing chrome devtools logic and have it mostly work. This would allow 1st class support for `flutter run -d firefox`

**Comment by jonahwilliams** [comment]
More info: this is only available on the nightly download and doesn't seem to be supported on windows.

**Comment by jonahwilliams** [comment]
more breadcrumbs: https://developer.mozilla.org/en-US/docs/Tools/Remote_Debugging/Debugging_Firefox_Desktop

**Comment by Zenb0t** [comment]
Any updates on this? Would be nice to test in a different browser...

**Comment by NicTanghe** [comment]
Flutter is ran by google highly doubt it will happen.

**Comment by jonahwilliams** [comment]
All the updates are provided above. in summary, it does not seem like it is possible to connect to the devtools of a Firefox browser unless you are using a nightly build, and even then only on certain platforms. Connecting to the browser devtools is a requirement for making Flutter devtools fully work.

We could build the Flutter tool to open firefox with the newly compiled app and support basic hot restart without any additional debugging functionality. This would mean that if you launched firefox from your IDE, you would be unable to do things like set breakpoints. Functionally, it would not be much different than flutter run -d web-server and opening up that page in Firefox, so we haven't spent any time on it. I'd also be worried it would be confusing to folks that are used to how Chrome is supported today.

@NicTanghe I resent the accusation that we are intentionally withholding functionality due to some vague implied conspiracy. Many aspects of Flutter are a labor of love and that includes Flutter Web. We have chosen to focus on performance and compatibility for the Web instead of investing significantly more time in expanding the debugging experience. I realize that not everyone will be satisfied with that - but the cost of working on a popular project is that there are many more requested features than we have time to work on.

If anyone wants to spend _their_ time working on this, you would certainly get no objections from me - and I'd be happy to help review patches and provide guidance. I just can't help when it comes to actually connecting to firefox, since that is the whole reason this isn't done.

**Comment by caniko** [comment]
This seems more like firefox issue than a flutter issue.

**Comment by NicTanghe** [comment]
I’m not saying you are withholding functionality.
I’m saying most of the work is being done by people that work at Google, and it would be contra productive to spend time on creating said functionality. 


"C:\Program Files\Mozilla Firefox\firefox.exe" -start-debugger-server

/Applications/Firefox.app/Contents/MacOS/firefox -start-debugger-server

firefox -start-debugger-server

Might fix a hurdle.

**Comment by pixiekat** [comment]
Yes, please. Firefox is my main browser and I hate having to have an instance of Chrome open just for one thing without all of my settings.

**Comment by Himura2la** [comment]
The Firefox discrimination is outrageous! I don't have Chrome because I purposefully avoid using Chromium-based browser. And everyone should! Please, support Firefox in `flutter doctor`, this is essential!

**Comment by leSs-cOws** [comment]
@Himura2la They also removed web server support. If you research a bit back, you used to be able to run `flutter run -d web-server` but you can't anymore for some reason.

**Comment by vladislav1nikov** [comment]
I think it will be good to add not only Firefox but also Safari. Chrome, Firefox and Safari are the most popular browsers and all use different engines. I think they are all important

**Comment by Ridhubharan** [comment]
Any updates?

**Comment by SakshamKarnawat** [comment]
Nearing 2024 but no support for Firefox is concerning. I love Flutter but please don't make me install Chrome 😟

**Comment by some1onLSD** [comment]
We still need this

**Comment by KatPurpy** [comment]
bump.

**Comment by pixiekat** [comment]
Either support Firefox or change `Chrome - develop for the web` to `Chrome - develop for Chrome`.

**Comment by juxuanu** [comment]
https://hacks.mozilla.org/2023/12/puppeteer-webdriver-bidi/

Could Flutter use webdriverbidi instead of the proprietary Chrome protocol, so that it could be used with any browser implementing this protocol?

**Comment by Thribs** [comment]
bump

**Comment by Ken-Andre** [comment]
4 years around now, and didn't do that. So not her priority. When we all think that only Apple being so restricted environment for devs; Now think that Google and all his open-sourced products web oriented, joined him(Apple).

**Comment by zamirszn** [comment]
its june aleady and no update on this yet

**Comment by rempas** [comment]
I'm glad I found this thread and I won't spend more time with Flutter! Let's be honest, we know this will never happen as Google wants you to use their browser. And what's even worse, is that it's the proprietary one. They don't at least let you use chromium. But yeah, how would they steal your data if they did?

I now understand why Flutter is not popular, despite been one of the very few frameworks that are stable and allow you to deploy to every target (except BSDs of course because, who cares about them, right?) without using a webkit. I will look at [Svelte](https://svelte.dev/) and [Tauri](https://tauri.app/) to build my apps in every platform! Yes, it uses webkits and will be slower but first, I don't care as I'll build commercial apps (that won't require extreme performance) and second, at least I don't have to deal with **BULLSHIT** and I can actually build an app, from tools made from an organization that actually cares about devs. 

Once again, thank you google from reminding us who you truly are behind the mask of the "open source lover"! Now, go ahead with server-side ads on YouTube, so, I also abandon it!

**Comment by provokateurin** [comment]
> They don't at least let you use chromium

You are wrong about this part: you can use `export CHROME_EXECUTABLE="chromium"`.

**Comment by deandreamatias** [comment]
> I'm glad I found this thread and I won't spend more time with Flutter! Let's be honest, we know this will never happen as Google wants you to use their browser. And what's even worse, is that it's the proprietary one. They don't at least let you use chromium. But yeah, how would they steal your data if they did?
> 
> I now understand why Flutter is not popular, despite been one of the very few frameworks that are stable and allow you to deploy to every target (except BSDs of course because, who cares about them, right?) without using a webkit. I will look at [Svelte](https://svelte.dev/) and [Tauri](https://tauri.app/) to build my apps in every platform! Yes, it uses webkits and will be slower but first, I don't care as I'll build commercial apps (that won't require extreme performance) and second, at least I don't have to deal with **BULLSHIT** and I can actually build an app, from tools made from an organization that actually cares about devs.
> 
> Once again, thank you google from reminding us who you truly are behind the mask of the "open source lover"! Now, go ahead with server-side ads on YouTube, so, I also abandon it!

This comment is a fair claim but won't change anything.
If want a another nice framework multipltaform, use Kotlin Multiplatform

**Comment by TechnicJelle** [comment]
> > They don't at least let you use chromium
> 
> You are wrong about this part: you can use `export CHROME_EXECUTABLE="chromium"`.

Exactly. In fact, I've been using Ungoogled Chromium for ages now.
It's been working great.
Though I would also prefer actual Firefox support...

**Comment by rempas** [comment]
> > They don't at least let you use chromium
> 
> You are wrong about this part: you can use `export CHROME_EXECUTABLE="chromium"`.

Thanks for the information! That is something at least! Tho, actual Firefox like @TechnicJelle or even better, any browser you want, would be the optimal choice.

**Comment by smunshi0mi** [comment]
This is incompetence, plain and simple. Wish you wouldn't ONLY focus on gulping free doughnuts.

**Comment by zs-dima** [comment]
Google has other priorities at the moment, take a look at Filip Hráček videous, plus it is the AI boom. And even for Flutter, metaprogramming and WASM are priorities.

**Comment by smunshi0mi** [comment]
@zs-dima It's been 4 years now. If they can rake in billions in profits, they can hire more people to work on the booms and the priorities. Yes, I can submit a pull request but so can Sundar Pichai, right?

**Comment by zs-dima** [comment]
@smunshi0mi 
1. They decide on their own how to waste their money. ?Unfortunately it is no way to force companies and individuals to spend money in a ways others want.
2. There are many other critical bugs waiting for more than 4 years.
https://github.com/flutter/flutter/issues/149031

**Comment by smunshi0mi** [comment]
@zs-dima

1.  Love how you defend a corporation with more dirt. 
2.  Two wrongs don't make a right.
3.  Don't care about your opinions.

**Comment by juxuanu** [comment]
https://hacks.mozilla.org/2024/08/puppeteer-support-for-firefox/

**Comment by bkonyi** [comment]
In order for us to consider supporting Firefox as a launch target, we need to be able to support debugging and Dart developer tooling.

Given that Mozilla has stated that they're deprecating support for the [Chrome DevTools protocol (CDP) EOY 2024](https://groups.google.com/a/mozilla.org/g/dev-platform/c/Z6Qu3ZT1MJ0), debugging support would require us to reimplement our web developer services to support a completely different protocol. This would be a significant engineering effort, requiring months of design and implementation work, so the Dart and Flutter teams are unlikely to prioritize this work.

**Comment by pixiekat** [comment]
> In order for us to consider supporting Firefox as a launch target, we need to be able to support debugging and Dart developer tooling.
> 
> Given that Mozilla has stated that they're deprecating support for the [Chrome DevTools protocol (CDP) EOY 2024](https://groups.google.com/a/mozilla.org/g/dev-platform/c/Z6Qu3ZT1MJ0), debugging support would require us to reimplement our web developer services to support a completely different protocol. This would be a significant engineering effort, requiring months of design and implementation work, so the Dart and Flutter teams are unlikely to prioritize this work.

Or, you know, Chrome could support open standards instead of proprietary DevTools.

**Comment by samirm** [comment]
@bkonyi would you be willing to provide a detailed plan of what needs to be considered/implemented to support a new protocol such as the one from FF? this way at least the community has a path forward to work on this if they really want it

**Comment by bkonyi** [comment]
@samirm, in all honesty, it would be a fairly hard sell to add support for an additional web browser debugging protocol as it would drastically increase the maintainable surface area of our web debugging services. It's unlikely that we'd have the bandwidth long term to take on the additional maintenance burden.

We do have some work ongoing to allow for our web debugging services to work without the Chrome DevTools protocol for limited use cases (e.g., no breakpoints or stepping, _maybe_ hot reload / hot restart support). If we get to a state where we can provide a reasonable amount of support for our developer tooling when deploying to Firefox, we might be able to reconsider this in the future.

We'd also like to explore making the `flutter` CLI tool extensible so the community can create custom deployment targets without having to ship them with the Flutter SDK. This is still in the early planning stages and may not happen, but we might be able to investigate this further in the next year.

**Comment by bkonyi** [comment]
Quick update for everyone here:

We've recently landed a change to enable limited developer tooling functionality for non-Chromium based browsers (e.g., Firefox and Safari), which includes hot reload and hot restart support. Functionality which requires communication with the browser's debugger (e.g., setting breakpoints, stepping through code, inspecting values, some Dart/Flutter DevTools tooling, etc.) won't work.

This feature is currently available on the `main` branch and will be part of the next stable release.

In general, we will continue to recommend using a Chromium based browser for the best Dart/Flutter for web development experience, but for those of you who are mostly interested in features like hot reload and widget inspector support, this should allow for that.

Currently, we don't have dedicated device targets for non-Chromium browsers, so the `web-server` device will need to be used. The Flutter team won't have cycles to add explicit support for common non-Chromium browsers as devices, but I would be happy to review well-tested PRs that add them.
